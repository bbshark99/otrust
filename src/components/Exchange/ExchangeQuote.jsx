import React, { useCallback, useEffect } from 'react';
import { BigNumber } from 'bignumber.js';
import ConfirmTransactionModal from 'components/Modals/components/ConfirmTransactionModal';
import PendingModal from 'components/Modals/components/PendingModal';
import ApproveTokensModal from 'components/Modals/components/ApproveTokensModal';
import RequestFailedModal from 'components/Modals/components/RequestFailedModal';
import TransactionCompletedModal from 'components/Modals/components/TransactionCompletedModal';
import TransactionFailedModal from 'components/Modals/components/TransactionFailedModal';
import { useWeb3React } from '@web3-react/core';

import { useChain } from 'context/chain/ChainContext';
import { BondingCont, NOMCont } from 'context/chain/contracts';
import { useExchange, useUpdateExchange } from 'context/exchange/ExchangeContext';
import {
  ExchangeItem,
  Sending,
  Receiving,
  ExchangeInput,
  MaxBtn,
  ReceivingValue,
  ExchangeButton,
} from './exchangeStyles';
import { useModal } from 'context/modal/ModalContext';
import NOMButton from 'components/Exchange/NOMButton';
import { format18, parse18 } from 'utils/math';

export default function ExchangeQuote({ strength }) {
  const { strongBalance, weakBalance } = useChain();
  const { handleModal } = useModal();
  const { library } = useWeb3React();

  const bondContract = BondingCont(library);
  const NOMcontract = NOMCont(library);

  const { askAmount, bidAmount, approveAmount, bidDenom, input, output, approve, strong, weak } = useExchange();
  const { NOMallowance } = useChain();

  useEffect(() => {
    console.log('Input: ', input);
    console.log('Output: ', output);
    console.log('Approve: ', approve);
  });

  const { objDispatch, strDispatch } = useUpdateExchange();

  const getAskAmount = useCallback(
    async (askAmountState, bidAmountUpdate, textStrength) => {
      var askAmountUpdate = askAmountState;

      switch (textStrength) {
        case 'strong':
          console.log('Strong: ', bidAmountUpdate.toFixed(0));
          askAmountUpdate = await bondContract.buyQuoteETH(bidAmountUpdate.toFixed(0));
          console.log('Pull Strong Ask Amount', askAmountUpdate);
          break;

        case 'weak':
          askAmountUpdate = await bondContract.sellQuoteNOM(bidAmountUpdate.toFixed(0));
          console.log('Pull Weak Ask Amount', askAmountUpdate);
          break;

        default:
          console.error('Denom not set');
      }
      return new BigNumber(askAmountUpdate.toString());
    },
    [bondContract],
  );

  const onApprove = async () => {
    if (weakBalance.gte(bidAmount)) {
      if (bidAmount.gt(NOMallowance)) {
        const approvalAmount = bidAmount.minus(NOMallowance);
        let objUpdate = new Map();
        objUpdate = objUpdate.set('approveAmount', approvalAmount);

        objDispatch({
          type: 'update',
          value: objUpdate,
        });

        let strUpdate = new Map();
        strUpdate = strUpdate.set('approve', format18(approvalAmount).toString());

        strDispatch({
          type: 'update',
          value: strUpdate,
        });

        handleModal(<ApproveTokensModal onConfirmApprove={onConfirmApprove} />);
      } else {
        handleModal(<ConfirmTransactionModal submitTrans={submitTrans} />);
      }
    } else {
      handleModal(<TransactionFailedModal error={`${weak} Balance too low`} />);
    }
  };

  const onConfirmApprove = () => {
    try {
      console.log('Approve Amount Before confirming: ', approve);
      handleModal(<ConfirmTransactionModal isApproving submitTrans={submitTrans} />);
    } catch (e) {
      handleModal(<TransactionFailedModal error={e.code + '\n' + e.message.slice(0, 80) + '...'} />);
    }
  };

  const submitTrans = async (isApproving, slippage, gasPrice) => {
    handleModal(<PendingModal isApproving={isApproving} />);

    if (isApproving) {
      if (!approveAmount) return;
      try {
        console.log('Approve Amount: ', approveAmount.toString());
        console.log('Approve Amount BG: ', approveAmount);
        console.log('Approve Amount; string', approve);
        console.log('Approve Amount: Bid', bidAmount.toString());

        let tx = await NOMcontract.increaseAllowance(bondContract.address, approveAmount.toFixed(8), {
          gasPrice: gasPrice.toFixed(0),
        });

        tx.wait().then(() => {
          handleModal(<TransactionCompletedModal tx={tx} />);
        });
      } catch (e) {
        console.log(e);
        handleModal(<TransactionFailedModal error={e.code + '\n' + e.message.slice(0, 80) + '...'} />);
      }
    } else {
      if (!bidAmount || !askAmount) return;
      try {
        let tx;
        let gasFee;

        switch (bidDenom) {
          case 'strong':
            // Preparing for many tokens / coins
            switch (strong) {
              case 'ETH':
                const gasFeeRaw = await bondContract.estimateGas.buyNOM(askAmount.toFixed(0), slippage.toFixed(0), {
                  value: bidAmount.toFixed(0),
                });

                gasFee = new BigNumber(gasFeeRaw.toString());

                const bidAmountUpdate = bidAmount.minus(gasFee.times(gasPrice));
                const askAmountUpdateRaw = await bondContract.buyQuoteETH(bidAmountUpdate.toFixed(0));
                const askAmountUpdate = new BigNumber(askAmountUpdateRaw.toString());

                tx = await bondContract.buyNOM(askAmountUpdate.toFixed(0), slippage.toFixed(0), {
                  value: bidAmountUpdate.toFixed(0),
                  gasPrice: gasPrice.toFixed(0),
                  gasLimit: gasFee.toFixed(0),
                });

                tx.wait().then(() => {
                  handleModal(<TransactionCompletedModal tx={tx} />);
                });

                break;
              default: {
              }
            }
            break;

          case 'weak':
            switch (weak) {
              case 'wNOM':
                tx = await bondContract.sellNOM(bidAmount.toFixed(0), askAmount.toFixed(0), slippage.toFixed(0), {
                  gasPrice: gasPrice.toFixed(0),
                });

                tx.wait().then(() => {
                  handleModal(<TransactionCompletedModal tx={tx} />);
                });
                break;
              default: {
              }
            }
            break;

          default:
            console.log();
        }
      } catch (e) {
        // eslint-disable-next-line no-console
        console.error(e.code, e.message.message);
        // alert(e.message)
        handleModal(<TransactionFailedModal error={e.code + '\n' + e.message.slice(0, 80) + '...'} />);
      }
    }
  };

  const onBid = () => {
    switch (true) {
      case bidDenom !== strength:
        handleModal(<RequestFailedModal error="Please enter amount" />);
        break;
      case strength === 'strong' && strongBalance.gte(bidAmount):
        handleModal(<ConfirmTransactionModal submitTrans={submitTrans} />);
        break;
      case strength === 'weak' && weakBalance.gte(bidAmount):
        handleModal(<ConfirmTransactionModal submitTrans={submitTrans} />);
        break;
      default:
        handleModal(<RequestFailedModal error="Insufficient funds" />);
    }
  };

  const onMax = async () => {
    let strUpdate = new Map();
    strUpdate.set('bidDenom', strength);
    let bidMaxValue = strength === 'strong' ? strongBalance : weakBalance;

    strUpdate.set('input', format18(bidMaxValue).toString());

    let askAmountUpdate;

    try {
      askAmountUpdate = await getAskAmount(askAmount, bidMaxValue, strength);
    } catch (err) {
      if (err) {
        handleModal(<RequestFailedModal error={err.error.message} />);
      }
    }

    let objUpdate = new Map();

    objUpdate = objUpdate.set('askAmount', askAmountUpdate);

    objUpdate = objUpdate.set('bidAmount', bidMaxValue);

    objDispatch({
      type: 'update',
      value: objUpdate,
    });

    strUpdate.set('output', format18(new BigNumber(askAmountUpdate.toString())).toFixed(8));

    strDispatch({
      type: 'update',
      value: strUpdate,
    });
  };

  const onTextChange = useCallback(
    async (evt, textStrength) => {
      evt.preventDefault();
      const floatRegExp = new RegExp(/(^(?=.+)(?:[1-9]\d*|0)?(?:\.\d+)?$)|(^\d\.$)/);
      console.log('Component Strength: ', strength);
      console.log('Text Strength: ', textStrength);
      console.log('Bid Denom: ', bidDenom);
      let strUpdate = new Map();
      switch (true) {
        case bidDenom === strength && input === evt.target.value.toString():
          break;
        case evt.target.value === '' || evt.target.value === '.':
          {
            let objUpdate = new Map();

            objUpdate = objUpdate.set('askAmount', new BigNumber(0));

            objUpdate = objUpdate.set('bidAmount', new BigNumber(0));

            objDispatch({
              type: 'update',
              value: objUpdate,
            });
          }

          strUpdate = strUpdate.set('bidDenom', strength);

          strUpdate = strUpdate.set('input', evt.target.value.toString());

          strUpdate = strUpdate.set('output', '');

          strDispatch({
            type: 'update',
            value: strUpdate,
          });

          break;
        case floatRegExp.test(evt.target.value.toString()):
          console.log('Input after test', evt.target.value);
          const bidAmountUpdate = parse18(new BigNumber(parseFloat(evt.target.value).toString()));

          const inputUpdate = evt.target.value.toString();

          if (bidDenom !== strength) {
            strUpdate = strUpdate.set('bidDenom', strength);
          }

          var askAmountUpdate;

          try {
            askAmountUpdate = await getAskAmount(askAmount, bidAmountUpdate, textStrength);
          } catch (err) {
            if (err) {
              console.log(err.error.message);
              handleModal(<RequestFailedModal error={err.error.message} />);
            }
            break;
          }

          let objUpdate = new Map();

          objUpdate = objUpdate.set('askAmount', new BigNumber(askAmountUpdate.toString()));

          objUpdate = objUpdate.set('bidAmount', bidAmountUpdate);

          objDispatch({
            type: 'update',
            value: objUpdate,
          });

          strUpdate = strUpdate.set('input', inputUpdate);

          strUpdate = strUpdate.set('output', format18(new BigNumber(askAmountUpdate.toString())).toFixed(8));

          strDispatch({
            type: 'update',
            value: strUpdate,
          });

          break;
        default:
          handleModal(<RequestFailedModal error="Please enter numbers only. Thank you!" />);
      }
    },
    [askAmount, bidDenom, getAskAmount, handleModal, input, objDispatch, strDispatch, strength],
  );

  return (
    <ExchangeItem>
      <strong>Bid {strength === 'strong' ? strong : weak}</strong>
      <Sending>
        <strong>I'm bidding</strong>
        <ExchangeInput
          type="text"
          onChange={evt => onTextChange(evt, strength)}
          value={bidDenom === strength ? input : ''}
          placeholder="0.00"
        />
        {strength === 'strong' ? strong : weak}
        <MaxBtn onClick={() => onMax()}>Max</MaxBtn>
      </Sending>
      <Receiving>
        <strong>I'm asking</strong>
        <ReceivingValue>
          {strength === bidDenom ? output : ''} {strength === 'strong' ? weak : strong}
        </ReceivingValue>
      </Receiving>
      {strength === 'strong' ? (
        bidDenom === 'weak' ? (
          <ExchangeButton>Input Value</ExchangeButton>
        ) : bidAmount.lte(strongBalance) ? (
          input === '' ? (
            <ExchangeButton>Input Value</ExchangeButton>
          ) : (
            <ExchangeButton onClick={onBid}>Buy {strength === 'strong' ? weak : strong}</ExchangeButton>
          )
        ) : (
          <ExchangeButton>Low {strong} Balance</ExchangeButton>
        )
      ) : (
        <NOMButton onBid={onBid} onApprove={onApprove} />
      )}
    </ExchangeItem>
  );
}
