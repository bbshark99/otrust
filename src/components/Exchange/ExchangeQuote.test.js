import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import BigNumber from 'bignumber.js';

import {
  ThemeWrapper,
  ExchangeContextWrapper,
  UpdateExchangeContextWrapper,
  ChainContextWrapper,
  renderWithContext,
  ModalContextWrapper,
} from 'utils/testing';
import ExchangeQuote from './ExchangeQuote';

const testProps = {
  strength: '10',
};

const testModalContext = {
  handleModal: jest.fn(),
};

const testUpdateExchangeContext = {
  objDispatch: jest.fn(),
  strDispatch: jest.fn(),
  setInputPending: false,
};

const testChainContext = {
  strongBalance: new BigNumber(0),
  weakBalance: new BigNumber(0),
};

const testExchangeContext = {
  askAmount: new BigNumber(0),
  bidAmount: new BigNumber(0),
  bidDenom: 'strong',
  status: 'Not Approved',
  strong: 'ETH',
  weak: 'NOM',
};

describe('Given an ExchangeQuote component', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithContext(ExchangeQuote);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when strong balance changed', () => {
    it('should updated the ask amount', async () => {
      const { getByTestId } = renderWithContext(ExchangeQuote);
      const receivingValue = getByTestId('exchange-weak-balance');

      fireEvent.change(getByTestId('exchange-strong-balance-input'), { target: { value: '1' } });

      await waitFor(() => expect(receivingValue).toHaveTextContent('ETH'));
    });
  });

  describe('when max button clicked', () => {
    it('should calculated max balace correctly', async () => {
      const { queryByTestId } = render(
        ChainContextWrapper(
          UpdateExchangeContextWrapper(
            ExchangeContextWrapper(
              ModalContextWrapper(ThemeWrapper(ExchangeQuote, testProps), testModalContext),
              testExchangeContext
            ),
            testUpdateExchangeContext
          ),
          testChainContext
        )
      );
      const receivingValue = queryByTestId('exchange-weak-balance');

      fireEvent.click(queryByTestId('max-value-button'));

      await waitFor(() => expect(receivingValue).toHaveTextContent('ETH'));
    });
  });
});
