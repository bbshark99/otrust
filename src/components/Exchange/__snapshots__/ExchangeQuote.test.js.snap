// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given an ExchangeQuote component when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-khIgEk kMWoji"
  >
    <strong>
      Sell NOM
    </strong>
    <div
      class="sc-hTRkXV bNKhVO"
    >
      <div
        class="sc-gSYDnn htDKdP"
        style="width: 100%; padding-right: 16px;"
      >
        <strong>
          I'm selling
        </strong>
        <input
          class="sc-eJocfa fsZdx"
          data-testid="exchange-strong-balance-input"
          placeholder="0.00"
          type="text"
          value=""
        />
      </div>
      <div
        class="sc-gSYDnn htDKdP"
      >
        NOM
        <button
          class="sc-iNiQyp iZyljO"
          data-testid="max-value-button"
        >
          Max
        </button>
      </div>
    </div>
    <div
      class="sc-laZMeE iTycLx"
    >
      <strong>
        You will receive
      </strong>
      <span
        class="sc-oeezt fJVdMx"
        data-testid="exchange-weak-balance"
      >
         ETH
      </span>
    </div>
    <button
      class="sc-lbVvki sc-gGLxEB eoDudu jFTELy"
    >
       Input Value 
    </button>
  </div>
</DocumentFragment>
`;
